_posts/2019-11-01-Crossaccount-Metrics-Forwarding.md

```diff
@@ -0,0 +1,420 @@
+
+Snippet 1 -
+```python
+def fetch_json_cloudwatch(
+    namespace,
+    dimensions,
+    cloudwatch_client,
+    current_time
+):
+    metric_data_queries = []
+    stats = [
+        'Average',
+        'Minimum',
+        'Maximum',
+        'Sum'
+    ]
+    for metric_name in metric_names:
+        for stat in stats:
+            metric_data_queries.append(
+                {
+                    'Id': 'request{}{}'.format(
+                        metric_name,
+                        stat
+                    ),
+                    'MetricStat': {
+                        'Metric': {
+                            'Namespace': namespace,
+                            'MetricName': metric_name,
+                            'Dimensions': dimensions
+                        },
+                        'Period': 60,
+                        'Stat': stat
+                    }
+                }
+            )
+
+    res = cloudwatch_client.get_metric_data(
+        MetricDataQueries=metric_data_queries,
+        StartTime=(current_time - 120),
+        EndTime=(current_time - 60)
+    )
+    return res
+```
+This is the function which is used to fetch CloudWatch metrics. It fetches metrics from the time period between 2 minute(StartTime) before and 1 minute before(EndTime) the time of lambda trigger. Different parameters of the function is described below.
+
+a) Namespace - It is namespace of the CloudWatch metrics which is 'AWS/ECS' in this specific case.
+
+b) Dimensions - The dimension(/identity) for the metrics. In this case, It constitutes name of the cluster(cluster_name) and different services(service_name).
+
+c) MetricName - The name of the metric. In this case, It is CPUUtilization and MemoryUtilization.
+
+d) current_time - Current time approximated to closest minute.
+
+e) cloudwatch_client - Boto3 client for CloudWatch which is used to fetch metrics. It is the Cloudwatch client of the source account in this case.
+
+f) stat - The CloudWatch statistic to return.
+
+Snippet 2 -
+```python
+role_arns = [
+    'PROD_CROSSACCOUNT_TRANSFER_ROLE_ARN',
+    'QA_CROSSACCOUNT_TRANSFER_ROLE_ARN',
+    'INT_CROSSACCOUNT_TRANSFER_ROLE_ARN'
+]
+
+for role_arn in role_arns:
+    response = client.assume_role(
+        RoleArn=os.environ[role_arn],
+        RoleSessionName='metricfetchsession'
+    )
+    session = boto3.Session(
+        aws_access_key_id=response['Credentials']['AccessKeyId'],
+        aws_secret_access_key=response['Credentials']['SecretAccessKey'],
+        aws_session_token=response['Credentials']['SessionToken']
+    )
+
+    source_cloudwatch_client = session.client('cloudwatch')
+```
+
+This piece of code facilitate read access to CloudWatch metrics of different accounts(source) from which metrics needs to be forwarded. It is accomplished using different role arns for cross-account access. It loops over role arn of source accounts to create sessions.     
```

Comment:
... **facilitates** ... CloudWatch metrics **in** different accounts(source) from which metrics **need** to be forwarded. It loops over **the role arns** of **the** source accounts to create sessions.